components/brave_new_tab_ui/components/search/SearchContext.tsx

```diff
@@ -0,0 +1,126 @@
+  private result: AutocompleteResult | undefined
+  private resultListeners: Array<(result?: AutocompleteResult) => void> = []
+  private selectionListeners: Array<(selection: OmniboxPopupSelection) => void> = []
+
+  constructor() {
+    omniboxController.setPage(this.receiver.$.bindNewPipeAndPassRemote())
+  }
+
+  addResultListener(listener: (result?: AutocompleteResult) => void) {
+    this.resultListeners.push(listener)
+    if (this.result) listener(this.result)
+  }
+
+  removeResultListener(listener: (result?: AutocompleteResult) => void) {
+    this.resultListeners = this.resultListeners.filter(r => r !== listener)
+  }
+
+  addSelectionListener(listener: (selection: OmniboxPopupSelection) => void) {
+    this.selectionListeners.push(listener)
+  }
+
+  removeSelectionListener(listener: (selection: OmniboxPopupSelection) => void) {
+    this.selectionListeners = this.selectionListeners.filter(s => s !== listener)
+  }
+
+  autocompleteResultChanged(result: AutocompleteResult) {
+    this.result = result;
+    for (const listener of this.resultListeners) listener(result)
+  }
+
+  updateSelection(selection: OmniboxPopupSelection) {
+    for (const listener of this.selectionListeners) listener(selection)
+  }
+}
+
+export const search = new SearchPage()
+
+export function SearchContext(props: React.PropsWithChildren<{}>) {
+  const [searchEngine, setSearchEngineInternal] = React.useState<SearchEngineInfo>()
+  const [query, setQuery] = React.useState('')
+  const { result: searchEngines = [] } = usePromise(() => searchEnginesPromise, [])
+  const filteredSearchEngines = React.useMemo(() => searchEngines.filter(isSearchEngineEnabled), [searchEngines])
+
+  const setSearchEngine = React.useCallback((engine: SearchEngineInfo | string) => {
+    if (typeof engine === 'string') {
+      engine = searchEngines.find(e => e.host === engine || e.keyword === engine)!
+    }
+
+    if (!engine) return
+
+    setDefaultSearchEngine(engine)
+    setSearchEngineInternal(engine)
+  }, [searchEngines]);
+
+  // When we receive search engines, use the first one as our keyword, if we don't have a match.
+  React.useEffect(() => {
+    if (!searchEngines.length) return
+
+    const match = filteredSearchEngines.find(s => s.host === getDefaultSearchEngine())
+      ?? searchEngines[0]
+    setSearchEngine(match)
+  }, [filteredSearchEngines])
+
+  // When the query changes, notify the browser side.
+  React.useEffect(() => {
+    if (query) {
+      const keywordQuery = `${searchEngine?.keyword} ${query}`
+      omniboxController.queryAutocomplete(stringToMojoString16(keywordQuery), false);
+    } else {
+      omniboxController.stopAutocomplete(true)
+    }
+  }, [query, searchEngine])
+
+  const context = React.useMemo(() => ({
+    searchEngine,
+    setSearchEngine,
+    query,
+    setQuery,
+    searchEngines,
+    filteredSearchEngines
+  }), [searchEngine, setSearchEngine, filteredSearchEngines, query, searchEngines])
```

Comment:
Out of curiosity, `setQuery` is missing here but it's immutable, so doesn't affect useMemo(), right?

Reply:
Exactly, yeah :smile: